### ISSUE TITLE
Suppress 'licenses is empty' warning from `gem build` for private gems

### ISSUE THREAD

**Issue by dgholz** [rubygems/rubygems#6724]
### Describe the problem as clearly as you can

RubyGems specification reference says to leave license empty for commercial/proprietary/all rights reserved gem, but `gem build` shows a warning:
```
WARNING:  licenses is empty, but is recommended.  Use a license identifier from
http://spdx.org/licenses or 'Nonstandard' for a nonstandard license.
```

The spec says
> Not specifying a license means all rights are reserved; others have no right to use the code for any purpose.

This exactly matches the requirement for the gem (an internal one). Is it possible to provide a value for `license` that means the same as not specifying a value? Or another way to avoid the warning while leaving `license` unspecified?

I didn't understand how `Nonstandard` should be interpreted from reading the specification, nor from the list on SPDX.

I see #3665 & #5253, they didn't help me understand how to:
1. Avoid the warning about a missing `license`
2. Note that I want all rights reserved in the same way as not specifying `license`

### Did you try upgrading RubyGems?

Yes, am on version `3.4.13`

### Post steps to reproduce the problem

1. Make a gemspec like
```
Gem::Specification.new do |s|
  s.name = 'pincushion'
  s.version = '0.1.0'
  s.authors = ['dgholz']
  s.summary = "A real gem for trying things out"
  s.homepage = "https://github.com/dgholz/pincushion"
  s.files = `git ls-files`.split("\n")
end
```
3. Run `gem build`

### Which command did you run?

`gem build`

### What were you expecting to happen?

The gem is built without warnings.

### What actually happened?

```
WARNING:  licenses is empty, but is recommended.  Use a license identifier from
http://spdx.org/licenses or 'Nonstandard' for a nonstandard license.
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: pincushion
  Version: 0.1.0
  File: pincushion-0.1.0.gem
```

### Run `gem env` and paste the output below

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.4.13
  - RUBY VERSION: 3.1.2 (2022-04-12 patchlevel 20) [arm64-darwin21]
  - INSTALLATION DIRECTORY: /Users/danielholz/.gem/ruby/3.1.2
  - USER INSTALLATION DIRECTORY: /Users/danielholz/.gem/ruby/3.1.0
  - RUBY EXECUTABLE: /Users/danielholz/.rubies/ruby-3.1.2/bin/ruby
  - GIT EXECUTABLE: /opt/homebrew/bin/git
  - EXECUTABLE DIRECTORY: /Users/danielholz/.gem/ruby/3.1.2/bin
  - SPEC CACHE DIRECTORY: /Users/danielholz/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/danielholz/.rubies/ruby-3.1.2/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - arm64-darwin-21
  - GEM PATHS:
     - /Users/danielholz/.gem/ruby/3.1.2
     - /Users/danielholz/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - :concurrent_downloads => 8
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/danielholz/.gem/ruby/3.1.2/bin
     - /Users/danielholz/.rubies/ruby-3.1.2/lib/ruby/gems/3.1.0/bin
     - /Users/danielholz/.rubies/ruby-3.1.2/bin
     - /usr/local/sbin
     - /opt/homebrew/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /opt/X11/bin
     - /Library/Apple/usr/bin
```

**Comment by pboling** [comment]
I too would love to have some clarity about how to handle this for private, internal, commercial gems.  For now I am putting "Nonstandard", but I have no idea what that really entails (IANAL).

**Comment by deivid-rodriguez** [comment]
I'd personally be fine with removing this warning. Happy to hear other opinions though.

**Comment by pboling** [comment]
To play devil's advocate with my own experience related above - 

I think the pressure applied on the community to specify a license, when automated, is hugely beneficial.

It may help us avoid the mess where purported open source code ends up being ambiguous, and potentially dangerous for many use cases. It can get adopted as a dependency of popular tools, and end up causing a ripple in space-time when it has to be summarily ripped out of every app in the world (ref the similar situation of the mime-types fiasco - not equivalent exactly - but you can imagine how a similar scenario could unfold if libraries begin being written and released without any warnings for lack of license).

It's a good and meaningful nudge to a developer - but it is problematic noise for the closed source use case.

Maybe there could be a gemspec directive that disables the warning, as was suggested in OP:

```
spec.license = nil
```

That could effectively turn off the warning, and be accepted as an explicit non-license.

**Comment by deivid-rodriguez** [comment]
Agreed!

Allowing to set it to `nil` is more explicit, and enables us to keep the warning and possibly explicitly mention the `nil` usage in the case of private gems.

**Comment by rubyFeedback** [comment]
> I think the pressure applied on the community to specify a license, when automated, is hugely beneficial.

I don't mind licence clarifications and I think all my gems have a licence stated in the .gemspec file. But note that this top-down view also generates more warnings in general, which can be annoying. For instance, I deliberately do not specificy an upper-range for the gem version, largely because I realised that gem then does not allow me to download my own gems, on the commandline, which I found hugely when not at home (I have to be on campus/university/restricted areas sometimes and I re-use my gems because they are useful basic building blocks to tap into other related functionality). So that was when I deliberately omitted the upper bounds check, yet gem itself keeps on generating tons of warnings that - to me - are now irrelevant and annoying. It's no big deal, I just ignore them, but for some of my gems it generates about 50 lines of warnings, which is just spammy. Basic point to make is that one should be really careful when considering "let's apply pressure via spam, this will be GREAT" top-down design. This is of course not the only reason I support dgholz's issue request here - the description makes sense to me and I can relate to it (even if it is not so important to me either way). I think at the least the documentation/specification should be improved, or adapted.

**Comment by yotsutose** [comment]
Hello, I would like to contribute to this good first issue.

Over the weekend, I took the time to understand this code and its surrounding context by actively engaging with it. However, there were several points in the PR comments that I didn't completely grasp, and I hope you could clarify them for me.

- Is the main objective simply to eliminate the warning?
    - This seems fairly straightforward.
    - However, I understood from the discussion that there's an explicit intent to permit `nil`.
- Or is there a more complex requirement?
    - For instance, starting from changes to `default_value` in `specification.rb`.
- Regarding the type check,
    - I modified the `.gemspec` file to observe the output firsthand.
    - I tried assigning `nil`, inserting an empty array, and not including any entry at all.
    - In all cases, the output was identical, and there were no type errors.
    - I realized that `default_value` in `specification.rb` is an empty array and, through the `||=`, it will never be `nil`.
    - I couldn't reproduce the circumstances that led to the type error. Could you explain how this occurred?

**Comment by deivid-rodriguez** [comment]
This was closed by #6879!
